### PR TITLE
Feat: detect changes also in configuration files

### PR DESCRIPTION
**Pull request by mikeleppane**
Detect changes also in configuration files. Detected config files are: pyproject.toml, ruff.toml,.ruff.toml

### CODE REVIEW

File: crates/ruff_cli/src/lib.rs [hunk 1]
```diff
@@ -213,13 +228,7 @@ fn check(args: CheckArgs, log_level: LogLevel) -> Result<ExitStatus> {
...
             match rx.recv() {
                 Ok(event) => {
-                    let paths = event?.paths;
-                    let py_changed = paths.iter().any(|path| {
-                        path.extension()
-                            .map(|ext| ext == "py" || ext == "pyi")
-                            .unwrap_or_default()
-                    });
-                    if py_changed {
+                    if has_file_changed(&event?.paths) {
```
Comment (MichaReiser) [inline crates/ruff_cli/src/lib.rs:231]: I think we have to recompute most of the variables defined above starting from line 115 whenever a `pyproject.toml` changes. For example, you can configure the output format in the settings and we should respect that change if we support reloading configuration files.

File: crates/ruff_cli/src/lib.rs [hunk 2]
```diff
@@ -43,6 +45,19 @@ impl From<ExitStatus> for ExitCode {
...
 
+pub fn has_file_changed(paths: &[PathBuf]) -> bool {
+    for path in paths {
+        if let (Some(file_name), Some(suffix)) = (path.file_name(), path.extension()) {
+            if let (Some(file_name), Some(suffix)) = (file_name.to_str(), suffix.to_str()) {
+                if WATCHED_FILES.contains(&file_name) || WATCHED_FILES.contains(&suffix) {
+                    return true;
+                }
+            }
+        }
```
Comment (MichaReiser) [inline crates/ruff_cli/src/lib.rs:56]: Nit: I recommend matching on the extension or file name explicitly. E.g. we don't want to match a file named `py`

```suggestion
	if matches!(path.file_name(), Some("pyproject.toml" | "ruff.toml" |".ruff.toml")) {
		true
	} else if matches!(path.extension(), Some("py" | "pyi")) { 
		true
	} else { false }
```

Comment (mikeleppane) [reply crates/ruff_cli/src/lib.rs:56]: Good catch!

File: crates/ruff_cli/src/lib.rs (hunk 1 above)
Comment (mikeleppane) [reply crates/ruff_cli/src/lib.rs:231]: Absolutely! Is there now something more to be recalculated other than pyproject_strategy?

Comment (MichaReiser) [reply crates/ruff_cli/src/lib.rs:231]: I'm not too familiar with that code but it is probably the `pyproject_strategy` and all data that gets derived from it (all settings that are supported in the configuration and aren't CLI options only).

File: crates/ruff_cli/src/lib.rs [hunk 3]
```diff
@@ -43,6 +43,42 @@ impl From<ExitStatus> for ExitCode {
...
 }
 
+#[derive(Default)]
+pub struct FileChangeDetector<'a> {
+    extensions: Vec<&'a str>,
+    config_files: Vec<&'a str>,
+    has_configuration_changed: bool,
+}
+
+impl FileChangeDetector<'_> {
```
Comment (MichaReiser) [inline crates/ruff_cli/src/lib.rs:53]: We can keep this a standalone function and instead introduce a new tri-boolean type to signal whether to ignore the change, it's a python file, or a configuration change

```rust
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
enum ChangeKind {
	Configuration,
	SourceFile,
	Ignore
}

fn detect_change(paths: &[PathBuf]) -> ChangeKind {
	...
}
```

File: crates/ruff_cli/src/lib.rs [hunk 4]
```diff
@@ -210,16 +246,19 @@ fn check(args: CheckArgs, log_level: LogLevel) -> Result<ExitStatus> {
...
-                    });
-                    if py_changed {
+                    if file_change_detector.change_detected(&event?.paths) {
+                        if file_change_detector.has_configuration_changed {
+                            pyproject_strategy = resolve::resolve(
+                                cli.isolated,
+                                cli.config.as_deref(),
+                                &overrides,
+                                cli.stdin_filename.as_deref(),
+                            )?;
```
Comment (MichaReiser) [inline crates/ruff_cli/src/lib.rs:260]: Only re-computing the `pyproject_strategy`, is, unfortunately, not sufficient. There are multiple derived variables on line 107-185 that need to be recomputed too

**Review by MichaReiser — CHANGES_REQUESTED** [review]

File: crates/ruff_cli/src/lib.rs (hunk 4 above)
Comment (charliermarsh) [reply crates/ruff_cli/src/lib.rs]: I added a TODO for this.

**Pull request merged**